### PR TITLE
Performance improvement: Reduce time spent reading settings file, variant 2

### PR DESCRIPTION
**Pull request by bjornhellander**
I profiled using StyleCopTester and noticed that a lot of time was spent reading the settings file. Previously a SourceTextValueProvider was used to cache the parsed json data, but that was no longer used. Updating SettingsHelper to again use a SourceTextValueProvider improved the time needed to retrieve the diagnostics in StyleCopTester by around 15% on my machine, when running on StyleCop.Analyzers.sln

This attempt uses RegisterCompilationStart in AnalysisContext to retrieve the settings file content early in the analysis process. Moved json file specific code from SettingsHelper to SettingsJsonHelper, to try to make the code easier to understand.

### CODE REVIEW

**Comment by bjornhellander** [comment]
Ready for review @sharwell. Let's see if this approach is better than #3519.

File: StyleCop.Analyzers/StyleCop.Analyzers/Settings/ValueOrException`1.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+namespace StyleCop.Analyzers
+{
+    using System;
+
+    /// <summary>
+    /// Stores the result of evaluating a function, either the return value
+    /// or any thrown exception.
+    /// </summary>
+    /// <typeparam name="T">The type of the returned value.</typeparam>
+    internal class ValueOrException<T>
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Settings/ValueOrException`1.cs:15]: 💡 `System.Lazy<T>` provides this same behavior

Comment (bjornhellander) [reply StyleCop.Analyzers/StyleCop.Analyzers/Settings/ValueOrException`1.cs:15]: Did not know that. Excellent!

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1602EnumerationItemsMustBeDocumented.cs [hunk 2]
```diff
@@ -63,7 +63,10 @@ public override void Initialize(AnalysisContext context)
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
             context.EnableConcurrentExecution();
 
-            context.RegisterSyntaxNodeAction(EnumMemberDeclarationAction, SyntaxKind.EnumMemberDeclaration);
+            context.RegisterCompilationStartAction(context =>
+            {
+                context.RegisterSyntaxNodeAction(EnumMemberDeclarationAction, SyntaxKind.EnumMemberDeclaration);
+            });
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1602EnumerationItemsMustBeDocumented.cs:69]: ❓  Was this change needed?

Comment (bjornhellander) [reply StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1602EnumerationItemsMustBeDocumented.cs:69]: Unsure if I misunderstood you, but HandleEnumMemberDeclaration takes a StyleCopSettings object as paremeter, if that was what you meant. Line 74.

**Comment by bjornhellander** [comment]
Closing since this is a mess and has conflicts
